### PR TITLE
Many new component solvers and autosolvers

### PR DESCRIPTION
**Pull request by eXish**
Added Component Solvers (TP Support) to the following:
- Binary Buttons
- Bridge
- Not Timer

Added Autosolvers for the following (most additions done using Shims):
- Accumulation
- Algebra
- Alphabet Numbers
- Audio Morse
- Blockbusters
- Boolean Keypad
- British Slang
- Catchphrase
- Christmas Presents
- Colored Keys
- Combination Lock
- Crazy Talk
- Cryptography
- The Cube
- Flashing Lights
- Hidden Colors
- Horrible Memory
- The Jack O' Lantern
- The Jukebox
- Listening
- Maintenance
- The Moon
- The Number Cipher
- Poker
- Probing
- Reverse Morse
- Simon's Star
- Snooker
- Sonic The Hedgehog
- Spinning Buttons
- The Stopwatch
- The Sun
- Symbolic Coordinates
- Tax Returns
- The Wire
- Wire Spaghetti
- Zoni

Fixed autosolvers solving too early for:
- Color Generator
- Hieroglyphics
- Lightspeed
- The Necronomicon
- The Sphere

Override the autosolvers of The Plunger Button and The Swan to fix bugs and make them more realistic.
Various code changes in existing Component Solvers to fix bugs.
Update Component Solver Factory with the new Component Solvers and Shims.

### CODE REVIEW

File: TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/SwanShim.cs [hunk 1]
```diff
@@ -23,6 +29,137 @@ protected override IEnumerator RespondToCommandShimmed(string inputCommand)
...
+			}
+		}
+		if (ans == ">: 77")
+		{
+			ans = ">: 7 7 ";
+			if (input == ">: 77")
+				input = ">: 7 7 ";
+		}
+		ans = ans.Replace(">:", "").Trim();
+		input = input.Replace(">:", "").Trim();
```
Comment (samfundev) [inline TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/SwanShim.cs:85]: Wouldn't it be better to do this to `input` when you get it initially and then not have `GetAnswer()` include `>:`?

**Review by samfundev — CHANGES_REQUESTED** [review]
Found a few things to fix up.

File: TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/ZoniShim.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+					case 1: correct = wordIndex == 0 || wordIndex == 8 || wordIndex == 13 || wordIndex == 23 || wordIndex == 26 || wordIndex == 37 || wordIndex == 48; break;
+					case 2: correct = wordIndex == 5 || wordIndex == 21 || wordIndex == 29 || wordIndex == 33 || wordIndex == 43; break;
+					case 3: correct = wordIndex == 9 || wordIndex == 12 || wordIndex == 28 || wordIndex == 35 || wordIndex == 51; break;
+					case 4: correct = wordIndex == 3 || wordIndex == 17 || wordIndex == 27 || wordIndex == 30 || wordIndex == 39 || wordIndex == 47 || wordIndex == 52; break;
+					case 5: correct = wordIndex == 1 || wordIndex == 22 || wordIndex == 41 || wordIndex == 50; break;
+					case 6: correct = wordIndex == 4 || wordIndex == 6 || wordIndex == 15 || wordIndex == 24 || wordIndex == 40 || wordIndex == 42; break;
+					case 7: correct = wordIndex == 2 || wordIndex == 7 || wordIndex == 19 || wordIndex == 36 || wordIndex == 44; break;
+					case 8: correct = wordIndex == 16 || wordIndex == 20 || wordIndex == 31 || wordIndex == 38 || wordIndex == 45 || wordIndex == 53; break;
+					case 9: correct = wordIndex == 11 || wordIndex == 25 || wordIndex == 32 || wordIndex == 46; break;
```
[the remark this answers is not in the document]
Comment (eXish) [reply TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/ZoniShim.cs:46]: Yeah sorry about this one, it was copy pasted from LeGeND's code.

File: TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/SwanShim.cs [hunk 3]
```diff
@@ -23,6 +29,137 @@ protected override IEnumerator RespondToCommandShimmed(string inputCommand)
...
+			goto reset;
+	}
+
+	string GetAnswer(int systemResetCounter, bool ready)
+	{
+		if (!ready)
+		{
+			return ">: 4 8 15 16 23 42 ";
+		}
+		else if (new[] { 8, 10, 14, 16, 24 }.Contains(systemResetCounter))
```
[the remark this answers is not in the document]
Comment (eXish) [reply TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/SwanShim.cs:118]: Yeah sorry again this was copy pasted from Royal.

File: TwitchPlaysAssembly/Src/ComponentSolvers/Helpers/ComponentSolverFactory.cs [hunk 4]
```diff
@@ -121,7 +123,7 @@ static ComponentSolverFactory()
 
 		//TheDarkSid3r Modules
 		ModComponentSolverCreators["AlienModule"] = module => new AlienFilingColorsComponentSolver(module);
-		ModComponentSolverCreators["QLModule"] = module => new QuiplashComponentSolver(module);
```
[the remark this answers is not in the document]
Comment (eXish) [reply TwitchPlaysAssembly/Src/ComponentSolvers/Helpers/ComponentSolverFactory.cs:124]: Uh I could have sworn I did, I can check again.

File: TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/AccumulationShim.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+		_keypadButtons = (KMSelectable[]) KeypadButtonsField.GetValue(_component);
+		_clearButton = (KMSelectable) ClearButtonField.GetValue(_component);
+		_submitButton = (KMSelectable) SubmitButtonField.GetValue(_component);
+	}
+
+	protected override IEnumerator RespondToCommandShimmed(string inputCommand)
+	{
+		IEnumerator command = RespondToCommandUnshimmed(inputCommand);
+		while (command.MoveNext())
+			yield return command.Current;
```
[the remark this answers is not in the document]
Comment (eXish) [reply TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/AccumulationShim.cs:22]: Well I can check, and if your right I'll have to change a lot of my shims lol.

File: TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/SwanShim.cs (hunk 1 above)
Comment (eXish) [reply TwitchPlaysAssembly/Src/ComponentSolvers/Modded/Shims/SwanShim.cs:85]: The autosolver for this caused me many problems, so I didn't care about optimization at some point. I can look at it. Thanks for the comments again.

**Pull request merged**